xformers/components/positional_embedding/sine.py

```diff
@@ -4,32 +4,45 @@
 # LICENSE file in the root directory of this source tree.
 
 
+# Silence Mypy errors in this file.
+# type: ignore
+
+from __future__ import annotations
+
 import math
+from typing import TypeVar
 
 import torch
+from pyre_extensions import Generic
+from typing_extensions import Literal as L
 
 from xformers.components.positional_embedding import (
     PositionEmbedding,
     PositionEmbeddingConfig,
     register_positional_embedding,
 )
 
+N = TypeVar("N", bound=int)
+DimModel = TypeVar("DimModel", bound=int)
+
 
 @register_positional_embedding("sine", PositionEmbeddingConfig)
-class SinePositionalEmbedding(PositionEmbedding):
-    def __init__(self, dim_model: int, *args, **kwargs):
+class SinePositionalEmbedding(PositionEmbedding, Generic[DimModel]):
+    def __init__(self, dim_model: DimModel, *args, **kwargs):
         super().__init__()
-        self.dim_model = dim_model
+        self.dim_model: DimModel = dim_model
 
-    def forward(self, x: torch.Tensor):
+    def forward(
+        self, x: torch.Tensor[torch.float, N, N]
+    ) -> torch.Tensor[torch.float, N, N, DimModel]:
         seq_len = x.shape[1]
         pos = (
-            torch.arange(0.0, seq_len, device=x.device)
+            torch.arange(0, seq_len, device=x.device, dtype=torch.float32)
```

Comment:
`torch.arange(0, seq_len, device=x.device, dtype=x.dtype)`, else this will error out if x is fp16

Reply:
oh, thinking about it, it's even strange that it works right now

Reply:
That's what I tried at first, but even the unit test fails because `x` can be `int`. Alternatives welcome. (Or is the unit test misleading?)

Reply:
oh of course, sorry, it has to be a float type (inputs are typically tokens, which are int), fp32 should be fine actually (the layers consuming this can always cast it down to fp16). All good